### PR TITLE
EZP-30562: Make it possible to ignore tags over configurable max length

### PR DESCRIPTION
**Pull request by ITernovtsiy**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30562](https://jira.ez.no/browse/EZP-30562)
| **Type**           |  Feature
| **Target version** | `0.9` / `master`
| **BC breaks**      | no
| **Doc needed**     | yes

related to 
https://github.com/ezsystems/ezplatform-http-cache/pull/86 
https://github.com/ezsystems/ezplatform-http-cache/issues/79
While using shortened tags will help to "reduce the risk" of 500 error, it still will be possible to embed too many objects in admin and got a 500 error page.

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests + specs and passing (`$ composer test`)
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

**Comment by ITernovtsiy** [comment]
@andrerom please review when you'll have a chance, thanks.

**Comment by andrerom** [comment]
Status: This is part of a bigger problem, we _also_ need to find a good way to reduce size of tags, and if they go above a certain size _(i.e. 8kb, which will cause issues several places)_, log warning and shorten.

I started some time ago to try to cleanup #86, however I now think there are two options to consider:
- Aim for new 0.10 release where all tags are shortened, to more or less what #86 proposes, no complexity trying to support both at the same time
- Like Drupal(?): Keep tags as is and use a hashing algo in production that produces shorter strings, strings doesn't need to be 100% unique, but as close to it as possible with as short strings as possible would be good. Can even have setting for this

--

So in summary: We can move forward on this, but as soon as we have smaller tags we should maybe do a follow up on this to set a sane default.

File: src/Handler/TagHandler.php [hunk 1]
```diff
@@ -57,19 +59,20 @@ public function tagResponse(Response $response, $replace = false)
         $tags = [];
         if (!$replace && $response->headers->has($this->tagsHeader)) {
             $headers = $response->headers->get($this->tagsHeader, null, false);
-            if (!empty($headers)) {
+            if ($headers) {
```
Comment (andrerom) [inline src/Handler/TagHandler.php]: irrelevant change?

File: src/Handler/TagHandler.php [hunk 2]
```diff
@@ -57,19 +59,20 @@ public function tagResponse(Response $response, $replace = false)
...
                 // handle both both comma (FOS) and space (this bundle/xkey/fastly) separated strings
                 // As there can be more requests going on, we don't add these to tag handler (ez-user-context-hash)
                 $tags = preg_split("/[\s,]+/", implode(' ', $headers));
             }
         }
 
         if ($this->hasTags()) {
+            // note that 'getTagsHeaderValue' returns unique tags
```
Comment (andrerom) [inline src/Handler/TagHandler.php]: yes, but after merge they are not guaranteed to be unique right?

File: src/Handler/TagHandler.php (hunk 1 above)
Comment (ITernovtsiy) [reply src/Handler/TagHandler.php]: yes, just a micro-optimization. I think it worsts it here, as this function called on every request. 
Let me know if I should remove it.

File: src/Handler/TagHandler.php (hunk 2 above)
Comment (ITernovtsiy) [reply src/Handler/TagHandler.php]: right, missed that, fixed

File: src/Handler/TagHandler.php (hunk 1 above)
Comment (andrerom) [reply src/Handler/TagHandler.php]: Then I think you should remove it.

Reason besides being irrelevant to the change here: afaik `empty` is faster, or _at least_ as fast as changing to evaluating if the value is true or false. `empty` like `isset` is language constructs, and not functions, so they don't have function overhead.

File: docs/using_tags.md [hunk 3]
```diff
@@ -172,4 +172,11 @@ One common issue to encounter is that the tagging headers exceed limits in Varni
 - [http_max_hdr](https://varnish-cache.org/docs/6.0/reference/varnishd.html#http-max-hdr) (e.g. 128)
 - [http_resp_size](https://varnish-cache.org/docs/6.0/reference/varnishd.html#http-resp-size) (e.g. 64k)
 
+In case you exceed the limit on Nginx, 500 error will be returned.
+Consider using fewer tags or ignore tags over the limit. It can be done by setting a max length parameter (in bytes)
+```yaml
+parameters:
+    ezplatform.http_cache.tags.header_max_length: 4000
```
Comment (andrerom) [inline docs/using_tags.md]:
```suggestion
    # Warning, setting this means you risk losing tag information, which means you can end up with stale cache
    # First tune your system if can, i.e. nginx: proxy_buffer_size, Varnish: http_resp_hdr_len, on Fastly limit is 16k
    ezplatform.http_cache.tags.header_max_length: 16000
```

File: src/Handler/TagHandler.php (hunk 1 above)
Comment (ITernovtsiy) [reply src/Handler/TagHandler.php]: Ok, reverted original code here and in `if (!empty($repoPrefix)) {`.

regarding `empty` vs `bool cast`, bool cast is faster (both with and without opcache optimization).
Proof:
empty_vs_boolcast
`!empty` variant performs 2 extra operations without opcache (just 1 extra after opcache optimization).

Comment (andrerom) [reply src/Handler/TagHandler.php]: I stand corrected :)

File: src/Handler/TagHandler.php [hunk 4]
```diff
@@ -80,7 +86,14 @@ static function ($tag) use ($repoPrefix) {
                 $tags[] = 'ez-all';
             }
 
-            $response->headers->set($this->tagsHeader, implode(' ', array_unique($tags)));
+            $tagsString = implode(' ', $tags);
+            if ($this->tagsHeaderMaxLength && strlen($tagsString) > $this->tagsHeaderMaxLength) {
+                $tagsString = trim(substr($tagsString, 0, strrpos(
```
Comment (andrerom) [inline src/Handler/TagHandler.php]: I'd like us to add a warning here when we end up dropping tags.

Comment (ITernovtsiy) [reply src/Handler/TagHandler.php]: Done, warning message:
HTTP Cache tags header max length reached and truncated to `$this->tagsHeaderMaxLength`
Let me know if it should be updated.

**Comment by ITernovtsiy** [comment]
@andrerom I can close this PR, right? I see this functionality released with v2.5.9

**Comment by andrerom** [comment]
Yes indeed, thanks :)
